User persistence with Doobie

dialelo opened the pull request:
~~Not ready for merge but opening for feedback, feel free to modify the list below~~:

- [x] Use Doobie instead of Slick for DB access
- [x] Dependency Injection with implicits
- [x] User operations algebra
- [x] User operations production interpreter
- [x] Better type signatures for user store & service
- [x] Better schema for users
 + [x] Evolution for the new schema 

Is not yet ready but any comments welcome, @raulraja also suggested lifting `ConnectionIO` to cats' Free monad which we can do in the future. Let me know your thoughts @rafaparadela @juanpedromoreno @andyscott.

juanpedromoreno commented on site/build.sbt:
```diff
@@ -62,7 +62,8 @@ lazy val server = (project in file("server"))
       "org.clapper" %% "classutil" % "1.0.5",
       "com.toddfast.typeconverter" % "typeconverter" % "1.0",
-      "org.scalaz" %% "scalaz-concurrent" % "7.2.0") ++
+      "org.scalaz" %% "scalaz-concurrent" % "7.1.2",
```
Maybe we could unify somehow the scalaz dependency version, since we are using different ones along the different modules: 7.1.2, 7.1.4, 7.2.0...

juanpedromoreno commented on site/server/app/com/fortysevendeg/exercises/ApplicationLoader.scala:
```diff
@@ -26,15 +33,15 @@ class Components(context: Context)
+  val applicationController = new ApplicationController
+  val exercisesController = new ExercisesController
+  val userController = new UserController
+  val OAuthController = new OAuth2Controller
```
IMHO `val` better if they are `camelCase` ;)

oauthController perhaps?

juanpedromoreno commented on site/server/app/com/fortysevendeg/exercises/controllers/ApplicationController.scala:
```diff
@@ -18,13 +18,14 @@ import play.api.mvc._
 
-class ApplicationController(userService: UserServices)(implicit exerciseOps: ExerciseOps[ExercisesApp]) extends Controller {
-
+class ApplicationController(implicit exerciseOps: ExerciseOps[ExercisesApp], userService: UserServices) extends Controller {
```
Line limit exceeded :)

juanpedromoreno commented on site/server/app/com/fortysevendeg/exercises/controllers/ApplicationController.scala:
```diff
@@ -35,8 +36,8 @@ class ApplicationController(userService: UserServices)(implicit exerciseOps: Exe
-          val user = userService.getUserByLogin(GetUserByLoginRequest(login = request.session.get("user").getOrElse("")))
-          user.map(response ⇒ Ok(views.html.templates.home.index(user = response.user, libraries = libraries)))
+          val user = userService.getUserByLogin(request.session.get("user").getOrElse(""))
+          Future.successful(Ok(views.html.templates.home.index(user = user, libraries = libraries)))
```
Minor question, if the user hasn't been found in the session... should we return `Future.successful`?

dialelo replied:
I think so, we can have anonymous users navigate the site with no problem. This won't allow exercise evaluation but the content is rendered so web crawlers can index it.

juanpedromoreno replied:
Makes sense :+1:

juanpedromoreno commented on site/server/app/com/fortysevendeg/exercises/controllers/ExercisesController.scala:
```diff
@@ -10,10 +10,11 @@ import scala.concurrent.Future
+import com.fortysevendeg.exercises.services.interpreters.ProdInterpreters._
 
-class ExercisesController(implicit exerciseOps: ExerciseOps[ExercisesApp]) extends Controller with JsonFormats {
+class ExercisesController(implicit exerciseOps: ExerciseOps[ExercisesApp], userServices: UserServices) extends Controller with JsonFormats {
```
Same here (long line ;) )

juanpedromoreno commented on site/server/app/com/fortysevendeg/exercises/services/UserServices.scala:
```diff
@@ -1,71 +1,79 @@
+    email:       String
+  ): Throwable Xor User = {
+    val user = User(None, login, name, github_id, picture_url, github_url, email)
+    Xor.fromOption(userStore.create(user), new Exception("Couldn't create user"))
```
I think we don't want generic exceptions. We could wrap this exception in a more specific one.

dialelo replied:
Definitely, I'm going to improve the error handling and type signatures.

juanpedromoreno commented:
Minor format comments added. I like this PR!! Good job!

dialelo commented:
I've made a few changes here, namely:

- Change APIs to avoid common pitfalls
 + Don't accept `User` as a parameter to update functions since its id might be `None`
 + Don't accept `User` as a parameter to create functions since ID's are provided by the DB
- Separate DB access in a couple of layers
 + `com.fortysevendeg.exercises.models.queries.Queries` constructs queries, which are typechecked in tests
 + `com.fortysevendeg.exercises.models.UserDoobieStore` constructs DB operations yielding `ConnectionIO` for composability, it doesn't implicitly require a transactor and doesn't run queries
 + `com.fortysevendeg.exercises.services.UserServiceImpl` takes an implicit transactor and runs operations in `UserDoobieStore`

I'd like to improve error handling but is difficult to do it in a DB-agnostic way. Since we are switching to Postgres soon my idea is to take advantage of Doobie's [Postgres error handling facilities](http://tpolecat.github.io/doobie-0.2.3/13-Extensions-PostgreSQL.html) for error handling to improve it, but that will be another PR.

raulraja commented on site/build.sbt:
```diff
@@ -27,6 +27,8 @@ lazy val exerciseSettings: Seq[Setting[_]] =
 
 lazy val clients = Seq(client)
+lazy val doobieVersion = "0.2.3"
+lazy val scalazVersion = "7.1.4"
```
scalaz-concurrent already brings in scalaz and so does doobie. We should try to align versions if they aren't already.

dialelo replied:
The "7.1.4" version is used, both for `scalaz-core` and `scalaz-concurrent`, ignoring the transitive dependency for doobie.

juanpedromoreno commented on site/server/app/com/fortysevendeg/exercises/controllers/UserController.scala:
```diff
@@ -6,24 +6,32 @@ import play.api.libs.json.Reads._
+import scalaz.{ -\/, \/, \/- }
 
-class UserController(userStore: UserStore) extends Controller {
+class UserController(implicit userOps: UserOps[ExercisesApp], userServices: UserServices) extends Controller {
```
Is `userServices` needed in this context?

dialelo replied:
Not at all, good catch!

raulraja commented on site/server/app/com/fortysevendeg/exercises/controllers/ApplicationController.scala:
```diff
@@ -35,8 +40,8 @@ class ApplicationController(userService: UserServices)(implicit exerciseOps: Exe
         request.session.get("oauth-token") map { token ⇒
-          val user = userService.getUserByLogin(GetUserByLoginRequest(login = request.session.get("user").getOrElse("")))
-          user.map(response ⇒ Ok(views.html.templates.home.index(user = response.user, libraries = libraries)))
+          val user = userService.getUserByLogin(request.session.get("user").getOrElse(""))
```
If `userService.getUserByLogin` returns an `Option` it means that it may be blocking. In which case the default executor pool may not be appropriate. This should all be in the context of `Free` and monadically computed along with `getLibraries` above so it can be lifted to the appropriate interpreter target monad.

raulraja replied:
Not saying we should do any of this now but if the UserService is not already lifting to Free or using Free Algebras we should refactor it or have a ticket to do so.

dialelo replied:
I'll try to address it as part of the proposed changes; if I see that is too involved I'll open an issue and work on it on a separate PR for making review easier.

juanpedromoreno commented on site/server/app/com/fortysevendeg/exercises/models/UserModel.scala:
```diff
@@ -20,81 +20,99 @@ trait UserStore {
+    Queries.byLogin(login).option.flatMap({
+      case None ⇒ sql"select false".query[Boolean].unique
+      case Some(u) ⇒ Queries.update(
+        u.id.get,
```
We shouldn't use `Option.get` never ;)

raulraja replied:
A for comprehension would be better suited for this case perhaps

juanpedromoreno commented on site/server/app/com/fortysevendeg/exercises/models/UserModel.scala:
```diff
@@ -20,81 +20,99 @@ trait UserStore {
-      } yield { db.run(users.filter(_.id === id).delete).map(_ == 1) }
-    }.map(!_.exists(i ⇒ !i))
+    Queries.update(
+      user.id.get,
```
Same here.

raulraja commented on site/server/app/com/fortysevendeg/exercises/models/UserModel.scala:
```diff
@@ -20,81 +20,99 @@ trait UserStore {
+object UserDoobieStore extends UserStore {
+  def all: ConnectionIO[List[User]] = {
+    Queries.all.list
   }
```
No need for {} here

raulraja commented on site/server/app/com/fortysevendeg/exercises/models/UserModel.scala:
```diff
@@ -20,81 +20,99 @@ trait UserStore {
-    db.run(users.sortBy(_.id.desc).result)
+  def getByLogin(login: String): ConnectionIO[Option[User]] = {
+    Queries.byLogin(login).option
   }
```
See comment above regarding {}

raulraja commented on site/server/app/com/fortysevendeg/exercises/models/UserModel.scala:
```diff
@@ -20,81 +20,99 @@ trait UserStore {
   }
 
+  def update(
+    id:          Int,
```
This should take a container instead of so many parameters. In general we favor single arg functions vs multiple ones.

dialelo replied:
What do you think about having an ADT with two constructors: `NewUser` and `User` instead of making the ID an `Option`? `NewUser` would have all the fields needed to create a user and `User`'s ID wouldn't be an `Option` anymore.

raulraja commented on site/server/app/com/fortysevendeg/exercises/services/interpreters/interpreters.scala:
```diff
@@ -25,15 +30,42 @@ object ProdInterpreters extends InterpreterInstances {
+    def apply[A](fa: UserOp[A]): Task[A] = fa match {
+      case GetUsers()            ⇒ Task.delay(userService.all)
+      case GetUserByLogin(login) ⇒ Task.delay(userService.getUserByLogin(login))
+      case CreateUser(
```
CreateUser should have been passed straight to the service as it may be decomposed there and applied to the query if needed via unapply

raulraja commented on site/server/app/com/fortysevendeg/exercises/utils/OAuth2.scala:
```diff
@@ -24,10 +25,12 @@ object OAuth2 {
 
   import OAuth2._
 
+  val ws = NingWSClient()
```
We need to look into the NingWSClient lifecycle in case it needs to be closed to avoid leaks like some clients have in the past.

raulraja commented on site/server/conf/evolutions/default/2.sql:
```diff
@@ -0,0 +1,16 @@
+    id integer primary key,
+    login varchar(255) unique not null,
+    name varchar(255) not null,
+    github_id varchar(255) not null,
```
We should try to avoid these _ leaking through Scala code and SQL if possible since it is not common practice in Scala to include such name conventions.

juanpedromoreno replied:
Agreed with @raulraja 

Particularly, I also like table names in UpperCase and using non-plural entity names:

For instance:

`CREATE TABLE User (... , githubId varchar(255), ...)`

This naming convention prevents things like this in our persistence model:

`val user1 = Users(1, "", "" , .. )`

But this a personal preference. What do you think guys @raulraja  @dialelo ?

dialelo replied:
:+1:

raulraja commented on site/server/test/Database.scala:
```diff
@@ -0,0 +1,32 @@
+  val testUsername = "sa"
+  val testPassword = ""
+
+  def db: Database = {
```
The name `db` it's misleading since it applies effects in applyEvolutions. Perhaps something like `updateDB()`

raulraja commented on site/server/test/UserStoreSpec.scala:
```diff
@@ -0,0 +1,103 @@
+    "create a new User and retrieve it afterwards" in {
+      val aUser = User(None, "47deg", "47 degrees", "47deg", "http://placekitten.com/50/50", "http://github.com/47deg", "hi@47deg.com")
+      val storedUser = UserDoobieStore.create(
+        aUser.login,
```
If User and CreateUser had similar shapes they could be automatically copied over

raulraja commented on site/shared/src/main/scala/shared/User.scala:
```diff
@@ -3,7 +3,7 @@ package shared
 
 case class User(
-  id:          Option[Long],
+  id:          Option[Int],
```
Why this change from Long to Int?

dialelo replied:
My bad, I'm going to change the `id`'s type in the database so we can map it to a `Long` in our code and have more available ID's than using integers.

juanpedromoreno commented on site/server/conf/evolutions/default/3.sql:
```diff
@@ -0,0 +1,16 @@
+# --- !Ups
```
If we are working in a PR in isolation maybe it makes sense to squash all **our** SQL changes in a single evolution. Do you agree?

dialelo replied:
Definitely, I'm going to squash all the evolutions into one with the final schema.

juanpedromoreno commented on site/server/test/UserStoreSpec.scala:
```diff
@@ -0,0 +1,103 @@
+
+  "UserDoobieStore" should {
+    "create a new User and retrieve it afterwards" in {
+      val aUser = User(None, "47deg", "47 degrees", "47deg", "http://placekitten.com/50/50", "http://github.com/47deg", "hi@47deg.com")
```
Same line in several places in this file. Could we extract this kind of code in factory method? We could receive all the parameters but with the less important ones with default values?

dialelo replied:
Great point.

dialelo commented:
I've addressed the comments in the PR and added a DB result algebra and implicit conversion that should allow us to run doobie transactions in our free for comprehensions. The `ExerciseServiceSpec` test fails but I can address that in a separate PR. 

Please take a look @raulraja  @juanpedromoreno @rafaparadela @andyscott :sparkles:

raulraja commented on site/server/app/com/fortysevendeg/exercises/app.scala:
```diff
@@ -1,27 +1,12 @@
-
-  }
-
+  type ExercisesAndUserOps[A] = Coproduct[ExerciseOp, UserOp, A]
```
I mentioned something similar in the PR from @juanpedromoreno. I don't believe the intermediate `Coproduct` aliases need to be named like these because they are not going to be used in isolation or composed in other context. Since they just represent aliases in the `Coproduct` tree constructions they may as well be named something like `C01`. If you disagree is fine, just a thought, since the only thing I need to know about those type aliases is in which order they are getting composed. It's similar to the use of the lambda name in type lambdas, this is just Scala being annoying in the lack of support for currying type constructors.

dialelo replied:
Nice, I'll change it to reflect that these are intermediate coproducts that shouldn't be referred to elsewhere, they are an implementation detail.

raulraja commented on site/server/app/com/fortysevendeg/exercises/controllers/ApplicationController.scala:
```diff
@@ -1,29 +1,34 @@
 package com.fortysevendeg.exercises.controllers
 
+import scala.language.postfixOps
```
We should add all these flags in the build.sbt as compiler flags. We will use infix notation and higher kinds elsewhere and there is no need to have these unnecesary advanced lang imports everywhere

raulraja commented on site/server/app/com/fortysevendeg/exercises/controllers/ApplicationController.scala:
```diff
@@ -1,29 +1,34 @@
+    implicit
+    exerciseOps: ExerciseOps[ExercisesApp],
+    userOps:     UserOps[ExercisesApp],
+    T:           Transactor[Task]
```
We should have cohesive naming for implicit evidences across the board.

raulraja commented on site/server/app/com/fortysevendeg/exercises/controllers/UserController.scala:
```diff
@@ -1,29 +1,44 @@
-  def index = Action { implicit request ⇒
-    Redirect("/")
+  def all = Action { implicit request ⇒
+    userOps.getUsers runTask match {
```
We should consider `runTask` to return an `Xor[A]` since we use cats elsewhere

dialelo replied:
Done.

raulraja commented on site/server/app/com/fortysevendeg/exercises/models/UserModel.scala:
```diff
@@ -1,100 +1,78 @@
+  def create(user: UserCreation.Request): ConnectionIO[UserCreation.Response] = for {
+    _ ← Queries.insert(user).run
+    user ← getByLogin(user.login)
+  } yield if (user.isDefined) Xor.Right(user.get) else Xor.Left(UserCreation.DuplicateName)
```
A `fold` or `pattern matching` is better suited for this case than manual runtime boolean check on presence / absence. It's literally safe the way it is but we don't want to encourage use of `Option.get` or any unsafe state accessors anywhere

raulraja commented on site/server/app/com/fortysevendeg/exercises/models/UserModel.scala:
```diff
@@ -1,100 +1,78 @@
+  def getOrCreate(user: UserCreation.Request): ConnectionIO[UserCreation.Response] = for {
+    maybeUser ← getByLogin(user.login)
+    theUser ← if (maybeUser.isDefined)
+      Xor.Right(maybeUser.get).point[ConnectionIO]
```
See comment above regarding the use of unsafe methods

raulraja commented on site/server/app/com/fortysevendeg/exercises/services/interpreters/interpreters.scala:
```diff
@@ -22,20 +42,54 @@ trait Interpreters[F[_]] extends InterpreterInstances {
-    def apply[A](fa: UserOp[A]): F[A] = ???
+  implicit def userOpsInterpreter(implicit A: ApplicativeError[F, Throwable], T: Transactor[F]): UserOp ~> F = new (UserOp ~> F) {
 
+    def apply[A](fa: UserOp[A]): F[A] = fa match {
```
`import UserDoobieStore._`

raulraja commented on site/server/app/com/fortysevendeg/exercises/services/interpreters/interpreters.scala:
```diff
@@ -48,12 +102,24 @@ trait InterpreterInstances {
+  implicit class FreeOps[A](f: Free[ExercisesApp, A]) {
+
+    /** Run this Free structure folding it's ops to an effect capturing task */
+    def runTask(implicit T: Transactor[Task]): Throwable \/ A = {
```
Let's return an `Xor` here so we don't leak the Scalaz types since we are using cats elsewhere

raulraja commented on site/server/app/com/fortysevendeg/exercises/utils/OAuth2.scala:
```diff
@@ -24,7 +30,11 @@ object OAuth2 {
-class OAuth2Controller(ws: WSClient, userService: UserServices) extends Controller {
+class OAuth2Controller(
+    implicit
+    transactor: Transactor[Task],
```
This is named `T` in other places, we should change all implicit evidences to have a cohesive naming convention.

dialelo replied:
In other places we needed the transactor implicit or otherwise the compiler would complain, I called it `T` because we were merely proving its existence. In this case I actually use the transactor instance so called it `transactor`, I'll rename it to `T` for consistency.

raulraja commented on site/server/app/com/fortysevendeg/exercises/utils/OAuth2.scala:
```diff
@@ -72,23 +82,24 @@ class OAuth2Controller(ws: WSClient, userService: UserServices) extends Controll
-          val github_id = (response.json \ "id").as[Long]
-          val avatar_url = (response.json \ "avatar_url").as[String]
-          val html_url = (response.json \ "html_url").as[String]
+          val githubId = (response.json \ "id").as[Long]
```
This is not safe but I understand we are changing this in another PR. All JSON parsing ops should be treated as an applicative builder style where a chunk is reduce to the structure or errors are reported accordingly, similar how the functional builders for JSON in play work.

dialelo replied:
I added a comment in #113 to address it in another PR.

raulraja commented on site/server/test/Database.scala:
```diff
@@ -0,0 +1,29 @@
+      url = testUrl,
+      config = Map("user" → testUsername, "password" → testPassword)
+    )
+  }
```
`{...}` Braces may be ommited

raulraja commented on site/server/test/UserQueriesSpec.scala:
```diff
@@ -0,0 +1,41 @@
+  check(Queries.deleteById(47))
+  check(Queries.update(
+    User(
+      47,
```
Use named parameters when params lists exceed 4-6 for readability

raulraja commented on site/server/test/UserStoreSpec.scala:
```diff
@@ -0,0 +1,90 @@
+
+  def before = UserDoobieStore.deleteAll.quick.run
+
+  def newUser(
```
By the way these params are applied this method may be not necessary since it's delegating all of them in order.

dialelo commented:
I've added property-based test for the user store.

dialelo commented:
Since we're not using Play's integration with Slick anymore evolutions aren't applied. Any ideas on how to make evolutions run automatically (at least in DEV mode)? @raulraja @juanpedromoreno

juanpedromoreno commented:
@dialelo @raulraja Let's create a new ticket to run the evolutions automatically with [Flyway](https://flywaydb.org/)

dialelo commented:
Created #161 to address schema evolutions.

raulraja commented on site/server/app/com/fortysevendeg/exercises/services/interpreters/interpreters.scala:
```diff
@@ -48,12 +102,27 @@ trait InterpreterInstances {
 /** Production based interpreters lifting ops to the effect capturing scalaz.concurrent.Task **/
-object ProdInterpreters extends Interpreters[Task]
+object ProdInterpreters extends Interpreters[Task] {
+  def scalazToCatsDisjunction[A, B](disj: \/[A, B]): Xor[A, B] =
```
This is cool but it could also be like...
```scala
def scalazToCatsDisjunction[A, B](disj: A \/ B): A Xor B
```

dialelo commented:
@raulraja just got rid of all the remaining calls to Option.get :fire:, take a look please!

raulraja commented:
Awesome! LGTM :+1: 
[![LGTM](http://lgtm.in/p/cgfWfFNnq)](http://lgtm.in/i/cgfWfFNnq)

dialelo commented:
:sparkles: :shipit: :sparkles:
